core/src/main/scala/org/apache/spark/rdd/HadoopRDD.scala

```diff
@@ -53,14 +53,22 @@ import org.apache.spark.util.{NextIterator, SerializableConfiguration, ShutdownH
 /**
  * A Spark split class that wraps around a Hadoop InputSplit.
  */
-private[spark] class HadoopPartition(rddId: Int, idx: Int, s: InputSplit)
+private[spark] class HadoopPartition(rddId: Int, override val index: Int, s: InputSplit)
   extends Partition {
 
   val inputSplit = new SerializableWritable[InputSplit](s)
 
-  override def hashCode(): Int = 41 * (41 + rddId) + idx
+  override def hashCode(): Int = 41 * (41 + rddId) + index
 
-  override val index: Int = idx
+  def canEqual(other: Any): Boolean = other.isInstanceOf[HadoopPartition]
+
+  override def equals(other: Any): Boolean = other match {
+    case that: HadoopPartition =>
+      super.equals(that) &&
+        (that canEqual this) &&
+        index == that.index
```

Comment:
Yes, but if changing the equality semantics is a different question, and it is, then we shouldn't change it here. I don't think it's better to add dummy implementations everywhere, since really, lacking `equals` is not wrong per se. It's lacking `hashCode` when `equals` has been defined that's incorrect and must be fixed/flagged.